pkg/gadgettracermanager/pubsub/pubsub.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pubsub
+
+import (
+	"sync"
+	"time"
+
+	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type PubSubEvent struct {
+	Deleted   bool
+	Container pb.ContainerDefinition
+	Wg        *sync.WaitGroup
+}
+
+// GadgetPubSub provides a synchronous publish subscribe mechanism for gadgets
+// to be informed of container creation and deletion. It needs to be
+// synchronous so that gadgets have time to attach their tracer before the
+// container is started.
+type GadgetPubSub struct {
+	mu sync.RWMutex
+
+	// subs contains a list of subscribers
+	subs []chan PubSubEvent
```

Comment:
Isn't it easier to have them in a map? Specially for the `Unsubscribe()` implementation?

Reply:
Yes, that would be easier for implementing Unsubscribe(). But for Publish(), we need to iterate over all channels. With a map, callers of Subscribe() would need to choose a name that does not collide with other names. Maybe the namespaced trace name would work?

Reply:
I think you can use the chan as map key and struct{} as value.